model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java

```diff
@@ -56,11 +61,22 @@ public static String getTableNameForNativeQuery(String tableName, EntityManager
         return (schema==null) ? tableName : dialect.openQuote() + schema + dialect.closeQuote() + "." + tableName;
     }
 
+    private static List<ParsedPersistenceXmlDescriptor> transformPersistenceUnits(Collection<PersistenceUnitDescriptor> descriptors) {
+        return descriptors.stream().map(descriptor -> (ParsedPersistenceXmlDescriptor) descriptor).collect(Collectors.toList());
+    }
+
     public static EntityManagerFactory createEntityManagerFactory(KeycloakSession session, String unitName, Map<String, Object> properties, boolean jta) {
         PersistenceUnitTransactionType txType = jta ? PersistenceUnitTransactionType.JTA : PersistenceUnitTransactionType.RESOURCE_LOCAL;
-        List<ParsedPersistenceXmlDescriptor> persistenceUnits = new ArrayList<>(PersistenceXmlParser.locatePersistenceUnits(properties));
+        PersistenceXmlParser parser = PersistenceXmlParser.create(properties);
+        List<URL> urls = parser.getClassLoaderService().locateResources("META-INF/persistence.xml");
 
-        persistenceUnits.add(PersistenceXmlParser.locateIndividualPersistenceUnit(JpaUtils.class.getClassLoader().getResource("default-persistence.xml")));
+        List<ParsedPersistenceXmlDescriptor> persistenceUnits = urls.isEmpty() ? new ArrayList<>() : transformPersistenceUnits(parser.parse(urls).values());
+        ParsedPersistenceXmlDescriptor defaultPersistenceUnit = transformPersistenceUnits(parser.parse(Collections.singletonList(JpaUtils.class.getClassLoader().getResource("default-persistence.xml")), txType)
```

Comment:
Wrap as a `singletonList` to be able to use the custom `NoSuchElementException` specified below. List.of() might potentially throw NPE, and we would not have the exact information about the failure.

I can share the reasoning if necessary.